Improve drain script support

jandubois opened the pull request:
This PR tries to address the remaining issues in the Quarks operator regarding drain script support.

Many thanks to @univ0298 for testing the changes and @andrew-edgar for updating the tests!

## History

The first identified issue was: [POD draining does not work because individual containers terminate prematurely instead of waiting for the completion of all drain/preStops on all containers](https://github.com/cloudfoundry-incubator/quarks-operator/issues/1297).

It was supposed to be fixed by: [Wait for all drain scripts to finish](https://github.com/cloudfoundry-incubator/quarks-operator/pull/1302)

That fix was incomplete because containers whose main process exits will get killed immediately and not participate in the drain script waiting process. The first idea how to fix this did not work out: [Make sure terminated containers are considered as already drained](https://github.com/cloudfoundry-incubator/quarks-operator/issues/1308)

Further problems are identified in: [Many Cloud Foundry drain scripts do not work in KubeCF due to no monit and no bpm](https://github.com/cloudfoundry-incubator/kubecf/issues/1728)

## Fixes

This pull request attempts to address the remaining issues with:

* keep `container-run` running until it receives a `SIGTERM` signal. That way `PreStop` hooks should not get killed prematurely.

* provide an emultation `bpm` script that implements the common `bpm stop` command used by many of the drain scripts in the CF releases. In addition it provides a mechanism to implement patches for those drain scripts that directly use the pid files to signal other processes.

In addition this PR fixes the way subprocesses are stopped via `SIGKILL`, orphaning their child processes and generally not allowing things to shut down gracefully. This PR mirrors the `bpm stop` functionality and sends a `SIGTERM`, followed 20s later with a `SIGKILL` if the child processes are still running at that time.

## BPM commands

The `bpm` script implements the following commands from the bpm-release:

```bash
bpm start job [-p process]
bpm stop job [-p process ]
```

The `bpm stop` command will wait up to 30s for the process to actually stop (it normally should be killed via `SIGKILL` within 20s if it doesn't respong to `SIGTERM`).

The following extensions are also provided:

```bash
bpm term job [-p process]
bpm quit job [-p process]
bpm running job [-p process]
```

* `bpm term` is the same as `bpm stop`, except it doesn't wait for the process to terminate.

* `bpm quit` will send a `SIGQUIT` signal to the process (e.g. to tell nginx to shut down gracefully).

* `bpm running` will return with a `0` exit code if the process is still running, or a `1` if it isn't. It will also print "yes" or "no" to stdout, if that is connected to a tty.

## Manual testing

I compiled the program on Linux (using `nc` with UDP to Unix sockets doesn't seem to work on macOS):

```
go build -o container-run cmd/main.go
```

I use the following test program for all tests:

```console
# cat main.sh 
#!/usr/bin/env bash

SIGQUIT=""
function quit {
    echo SIGQUIT
    SIGQUIT=1
}
trap quit SIGQUIT

function term {
    echo SIGTERM
    if [ -n "$SIGQUIT" ]; then
        echo "Exiting because already received SIGQUIT"
        exit 0
    fi
}
trap term SIGTERM

echo Start
for i in $(seq 10); do
    echo Loop $i
    sleep 3
done
echo End
exit 0
```

It runs for 30s and then exits on its own, printing a progress message every 3 seconds. It acknowledges `SIGQUIT` and `SIGTERM` signals, but only exits early if the `SIGTERM` has been preceeded by a `SIGQUIT`. It is always run via:

```
# ./container-run --job-name main --process-name main $PWD/main.sh; echo $?
Start
Loop 1
[...]
Loop 10
End
```

The `main.sh` process runs until completion, but `container-run` remains running. From a second shell `main.sh` can be restarted:

```console
# alias bpm=/var/vcap/jobs/bpm/bin/bpm
# bpm start main
```

and the process runs again from start to finish. Running `pkill container-run` once the script has finished will terminate `container-run` immediately now.

Starting `container-run` again and running `bpm stop main` from the second terminal shows that the signal is received (once the current sleep returns), but the process doesn't abort right away; it runs for another 20s before being killed:

```console
# time bpm stop main

real    0m20.040s
```

Output:

```console
# ./container-run --job-name main --process-name main $PWD/main.sh; echo $?
Start
Loop 1
Loop 2
SIGTERM
Loop 3
Loop 4
Loop 5
Loop 6
Loop 7
Loop 8
```

Using `bpm term main` works the same as `bpm stop main`, but returns immediately and doesn't wait for the process to stop.

Sending a `SIGQUIT` before the `stop` command lets the script terminate right away, but `container-run` will keep running as expected:

```console
# bpm start main; sleep 3; bpm quit main; bpm stop main
```

Output:

```console
[...]
Loop 8
Start
Loop 1
Loop 2
SIGQUIT
SIGTERM
Exiting because already received SIGQUIT
```

Sending `SIGTERM` to `container-run` while `main.sh` is running will send `SIGTERM` to the script as well. It will continue running because it hasn't received a `SIGQUIT` first. After 20s the script will be killed via `SIGKILL` and `container-run` exits:

```console
# bpm start main; sleep 3; pkill container-run
```

Output:

```console
[...]
Exiting because already received SIGQUIT
Start
Loop 1
Loop 2
SIGTERM
Loop 3
Loop 4
Loop 5
Loop 6
Loop 7
Loop 8
0
#
```

The `bpm running` command can be used to check if the process is still running or not:

```console
# bpm start main; bpm term main; while bpm running main; do sleep 5; done
yes
yes
yes
yes
no
```

## Open issues

In addition to a new operator release using the updated `container-run` version we also need at least 2 patches at the kubecf level:

* The cloud controller [drain.sh](https://github.com/cloudfoundry/capi-release/blob/develop/jobs/cloud_controller_ng/templates/drain.sh.erb) script calls [shutdown_drain.rb](https://github.com/cloudfoundry/capi-release/blob/develop/jobs/cloud_controller_ng/templates/shutdown_drain.rb.erb) using [cloud_controller_ng/drain.rb](https://github.com/cloudfoundry/cloud_controller_ng/blob/main/lib/cloud_controller/drain.rb). It should be possible to replace the drain script with something like:

    ```bash
    bpm quit nginx
    for i in $(seq 30); do
        test bpm running nginx && sleep 1
    done
    bpm term nginx
    bpm term cloud_controller_ng
    ```

* The garden process invokes `garden_start` via [tini](https://github.com/cloudfoundry/garden-runc-release/blob/develop/jobs/garden/templates/config/bpm.yml.erb#L2-L6). Sine we don't run `tini` using pid 1, we need to pass the `-s` option to enable the `TINI_SUBREAPER`.

jandubois committed:
Unregister subprocesses once they are no longer running

Signed-off-by: Jan Dubois <jan.dubois@suse.com>

jandubois committed:
Don't exit container-run until SIGTERM is received, or the main process fails

To avoid terminating early and killing a still running drain script it is
necessary to keep the container-run process running even if the main
process has already exited.

Once container-run receives the SIGTERM signal it will pass it on to the
child processes and then wait until all direct children have exited before
exiting itself. This should allow the processes to terminate cleanly as
long as the grace period has not yet expired.

container-run will also terminate immediately if the main process terminates
with an error. This allows the container to fail and be restarted by k8s.

Signed-off-by: Jan Dubois <jan.dubois@suse.com>

jandubois committed:
Add QUIT command to send SIGQUIT to process

This is used to trigger graceful shutdown of nginx.

Signed-off-by: Jan Dubois <jan.dubois@suse.com>

jandubois committed:
Create status indicator on the shared volume while the process is running

Can be used to wait for a process to stop after sending the STOP command.

Signed-off-by: Jan Dubois <jan.dubois@suse.com>

jandubois committed:
Write /var/vcap/jobs/bpm/bin/bpm script for drain script usage

It only supports the "start" and "stop" commands from the real bpm,
but has additional "quit" and "running" commands to help reimplementing
drain scripts that make use of the pid files.

Signed-off-by: Jan Dubois <jan.dubois@suse.com>

jandubois committed:
Add --debug option

Signed-off-by: Jan Dubois <jan.dubois@suse.com>

jandubois committed:
Make errors channel unbuffered again

It is important to process errors synchronously because their handling
depends on the value of the 'active' flag. To avoid deadlocks any code
sending errors to the channel must run in a blockable goroutine.

Signed-off-by: Jan Dubois <jan.dubois@suse.com>

jandubois committed:
Use SIGTERM instead of SIGKILL to STOP processes

(and give them a chance to notify their child processes). If
the processes haven't terminated in 20s, send SIGKILL.

Signed-off-by: Jan Dubois <jan.dubois@suse.com>

jandubois committed:
Add some additional debug logging

Signed-off-by: Jan Dubois <jan.dubois@suse.com>

jandubois committed:
Unregister process as soon as it quits

Signed-off-by: Jan Dubois <jan.dubois@suse.com>

jandubois committed:
Call stopProcesses on SIGTERM (and SIGINT for testing)

That way it also starts the timeout trigger if the process
doesn't stop within 20s.

Signed-off-by: Jan Dubois <jan.dubois@suse.com>

jandubois committed:
Only handle SIGINT, SIGQUIT, and SIGTERM signals

The rest are just confusing the debug output, and passing
on e.g. SIGCHLD to the child processes is neither correct
nor useful.

Signed-off-by: Jan Dubois <jan.dubois@suse.com>

jandubois committed:
bpm stop now waits up to 30s for the process to stop

container-run should send a SIGKILL if the process is still alive
after 20s, so if it isn't gone after 30, then it probably will not
stop.

The `bpm running` subcommand will also now print "yes" or "no" in
addition to setting the exit code.

Signed-off-by: Jan Dubois <jan.dubois@suse.com>

jandubois committed:
Make the tests compile again

(but they no longer pass because they have not been updated
with the new application logic).

Signed-off-by: Jan Dubois <jan.dubois@suse.com>

jandubois committed:
Add `bpm term` command; same as stop, but don't wait

Also modify `bpm running` to only print yes/no when stdout is a tty.

Signed-off-by: Jan Dubois <jan.dubois@suse.com>

andrew-edgar committed:
Refactor tests to work when Run does not exit

mook-as commented on pkg/containerrun/bpm.go:
```diff
@@ -0,0 +1,72 @@
+func WriteBPMscript() error {
+	fileName := "/var/vcap/jobs/bpm/bin/bpm"
+
+	script := `#!/bin/bash
```
It would be nice to put this in an actual file instead. But I'm not sure I want to deal with this code _enough_ to go through the hassle of doing it… In particular [embed](https://golang.org/doc/go1.16#embed) would be nice here, but I don't know if we have golang 1.16.

jandubois replied:
Github Actions try to build with Go versions back to 1.13: `go: [ '~1.13', '~1.14', '~1.15', '~1' ]`.

Also inlining scripts follows quarks-operator precedent: https://github.com/cloudfoundry-incubator/quarks-operator/blob/master/pkg/bosh/bpmconverter/containers.go#L308-L342

I think it is fine as-is.

mook-as commented on pkg/containerrun/bpm.go:
```diff
@@ -0,0 +1,72 @@
+else
+    # "term" is the same as "stop", except we won't wait
+    ACTION="${CMD/term/stop}"
+    echo "${ACTION^^}" | nc -w 1 -uU "${CONTAINER_RUN}.sock"
```
For my understanding only, no need to put this in the file: netcat (BSD), wait 1 second, UDP, Unix socket (path).  No long options available, as we're not using the GNU variant.

We simply echo `start`, `stop`, or `quit` to the pipe.

jandubois replied:
Yes, for whatever reason `nc` is the BSD version on SUSE.

"wait" meaning: timeout if a connection cannot be established within 1 second.

And the commands are uppercase: `START`, `STOP`, and `QUIT`.

mook-as replied:
Well, we have both versions available (`gnu-netcat` and `netcat-openbsd`), we just happen to [ask for the BSD variant](https://build.opensuse.org/package/view_file/Cloud:Platform:Fissile-Stemcell/fissile-stemcell/sle15-fissile-base.kiwi?expand=1).

jandubois commented:
> * we need to pass the `-s` option to enable the `TINI_SUBREAPER`

https://github.com/cloudfoundry-incubator/kubecf/pull/1729

jandubois commented:
> * The cloud controller [drain.sh](https://github.com/cloudfoundry/capi-release/blob/develop/jobs/cloud_controller_ng/templates/drain.sh.erb) script calls [shutdown_drain.rb](https://github.com/cloudfoundry/capi-release/blob/develop/jobs/cloud_controller_ng/templates/shutdown_drain.rb.erb) using [cloud_controller_ng/drain.rb](https://github.com/cloudfoundry/cloud_controller_ng/blob/main/lib/cloud_controller/drain.rb).

This new nginx drain logic was added in capi-release v1.110.0 via https://github.com/cloudfoundry/capi-release/commit/30690f303a5659249f1f202a4d88e9f3c0dc9729.

kubecf is currently using capi-release v1.107.0, so this does not yet apply.

mook-as commented on pkg/containerrun/containerrun.go:
```diff
@@ -222,20 +270,35 @@ func handlePacket(
-	for _, err := range processRegistry.SignalAll(os.Kill) {
-		errors <- err
-	}
+	var wg sync.WaitGroup
```
Hmm, why is a `WaitGroup` needed here? You never add to it, and it returns when the goroutine returns — which is exactly what would happen if everything was just inlined?

jandubois replied:
The difference is that the code writes to the `errors` channel, which is **unbuffered**, so the write does not return until the value has been read. Which means the write should not happen in the main goroutine.

Let me know if I misunderstand the situation, or if there is a better way to deal with it. Otherwise I would just add a comment.

jandubois replied:
Oops, forgot to add: the error handling needs to be synchronous because it depends on the setting of the `active` boolean in the main loop. I did make the channel buffered at first, but then realized that that won't work because there is a (minimal) state machine involved, so the sequencing matters.

mook-as replied:
The current thread (whatever entered `stopProcesses`) will block until `wg.Wait()` returns anyway, right? So it's not like adding the `WaitGroup` made things more concurrent.  The blocking write to the `errors` channel should also release the CPU for other threads to work, if I understand things correctly…

jandubois replied:
You are right, the `wg.Wait()` call negates the whole point of moving the code into a goroutine.

I still think the goroutine is necessary though: when the `Run()` function receives a `STOP` command it will call `stopProcesses`. `stopProcesses` will block if it writes to the `errors` channel, and since the reader for the channel is in the same loop in `Run()` that dispatched the `STOP` command it would deadlock.

So just leaving the `go func` in, but removing the `WaitGroup` should be the proper fix, right?

mook-as replied:
Looking at the callers [[1]] [[2]], [[1]] is basically asynchronous and we should return as fast as possible.  In [[2]] we land in the [big loop](https://github.com/cloudfoundry-incubator/quarks-container-run/pull/10/files#diff-3d94c8020c09ed9fc87868a6d076b1ad4dc4d85e2c0d0d07949009a3517686aeR206) and wait for the process count to hit zero before doing anything.  So yeah, I think we're safe; we just need to remove the extraneous wait group.

It might be useful to document this behaviour, possibly as the doc comment for the parameter for `HandleSignals`?

[1]: https://github.com/cloudfoundry-incubator/quarks-container-run/pull/10/files#diff-3d94c8020c09ed9fc87868a6d076b1ad4dc4d85e2c0d0d07949009a3517686aeR173
[2]: https://github.com/cloudfoundry-incubator/quarks-container-run/pull/10/files#diff-3d94c8020c09ed9fc87868a6d076b1ad4dc4d85e2c0d0d07949009a3517686aeR666

jandubois replied:
`HandleSignals` is already running in its own goroutine. So I just made the other call to `stopProcesses` invoke a goroutine at the call-site instead and removed all of the goroutine code from `stopProcesses`.

mook-as commented on pkg/containerrun/containerrun.go:
```diff
@@ -285,11 +352,21 @@ func startMainProcess(
 	}
 	processRegistry.Register(process)
 
+	sentinel := fmt.Sprintf("/var/vcap/data/%s/%s_containerrun.running", jobName, processName)
```
Do we need to worry about job and process names being weird? (For example, includes `/`)

jandubois replied:
I don't think so; these are the bosh labels, not the actual command name to start the process.

Also just copies the existing code for the socket name, just using a different extension: https://github.com/cloudfoundry-incubator/quarks-container-run/blob/master/pkg/containerrun/containerrun.go#L186

mook-as commented on pkg/containerrun/containerrun.go:
```diff
@@ -222,20 +270,35 @@ func handlePacket(
+		// but there doesn't seem to be a point to be doing it in this context.
+		processRegistry.timer = time.AfterFunc(sigtermTimeout, func() {
+			log.Debugln("timeout SIGTERM")
+			processRegistry.KillAll()
```
(Note that the `WaitGroup` is done before this goroutine runs)

jandubois replied:
Yes, that is fine. `KillAll` will just send a `SIGTERM` to any process still registered, and intentionally doesn't do any error handling.

mook-as commented on pkg/containerrun/containerrun.go:
```diff
@@ -504,28 +582,58 @@ type Stdio struct {
 	return len(pr.processes)
 }
 
+// Unregister removes a process from the registry and returns how many processes are registered.
```
```suggestion
// Unregister removes a process from the registry and returns how many processes are still registered.
```

mook-as commented on pkg/containerrun/containerrun.go:
```diff
@@ -504,28 +582,58 @@ type Stdio struct {
 
+// Count returns the number of processes in the registry
+func (pr *ProcessRegistry) Count() int {
+	return len(pr.processes)
```
Does this need to take a lock? Not that we care about the number that much, but mostly to ensure the race detector never makes a peep?

jandubois replied:
I assumed it was just reading a single integer, which should always be atomic, but adding the lock won't hurt, so I'll just do it.

mook-as commented on pkg/containerrun/containerrun.go:
```diff
@@ -504,28 +582,58 @@ type Stdio struct {
+	defer pr.Unlock()
+
+	log.Debugf("Unregistering process %s\n", p)
+	processes := make([]Process, 0)
```
```suggestion
	processes := make([]Process, 0, len(pr.processes))
```
(Making it the same size in case `p` is not in `pr.processes`; since this is `len()` not capacity, this will at most be one extra).

jandubois replied:
Can do, but doesn't really make a differences given that there will be 0, 1, or 2 processes in the array.

mook-as reviewed (changes requested):
Not really sure my review is comprehensive, but it's _something_ at least I guess?

jandubois committed:
Code review feedback

Signed-off-by: Jan Dubois <jan.dubois@suse.com>

jandubois committed:
Run stopProcesses in a goroutine when called from the main goroutine

The errors channel is unbuffered, so stopProcesses cannot deliver
errors if it is running from the same goroutine as the receiver.

Signed-off-by: Jan Dubois <jan.dubois@suse.com>

jandubois committed:
Fix indentation

Signed-off-by: Jan Dubois <jan.dubois@suse.com>

manno commented:
That looks great.

One thing I can think of, if any of the goroutines in the tests gives a problem (flakieness, etc) maybe `defer GinkgoRecover()` helps.
